### PR TITLE
Adapter v2

### PR DESCRIPTION
**Pull request by rasbt**
Adds adapter_v2 scripts for text (no multimodality)

- [x] add trainable bias and scale parameters
- [x] make new evaluation script
- [x] make new generation script
- [x] make normalization layers trainable
- [x] modify documentation
- [x] change back hparams for consistency with regular adapter

### CODE REVIEW

**Comment by rasbt** [comment]
Not sure if necessary, but I added simple test to check whether the model contains the modified parameters. 

Btw. I ran the finetuning and double-checked that the new parameters are updating during that. 

So far as I can tell, this PR should be ready.

File: lit_llama/adapter_v2.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+from lit_llama.adapter import LLaMA
+
+
+def adapter_v2_mark_only_adapter_as_trainable(model: LLaMA) -> None:
+    """Sets `requires_grad=False` for all non-adapter weights."""
+    for name, param in model.named_parameters():
+        substrings = ("adapter_wte", "gating_factor", "adapter_scale", "adapter_bias", "RMSNorm")
+        if "RMSNorm" in name:
+            print(name)
```
Comment (awaelchli) [inline lit_llama/adapter_v2.py:19]: The named parameters are always of the form `module.submodule.subsubmodule....`. I think you confused it with the class name for RMSNorm. In order to mark the rms norm trainable, we would have to include

```
("rms", "ln_f") 
```
or we could be a bit more strict by explicitly naming

```
("rms_1", "rms_2", "ln_f") 
```

File: finetune/adapter_v2.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    checkpoint = torch.load(pretrained_path)
+
+    with fabric.init_module():
+        model = LLaMA(config)
+        # strict=False because missing keys due to adapter weights not contained in state dict
+        model.load_state_dict(checkpoint, strict=False)
+
+        for module in model.modules():
+            if isinstance(module, nn.Linear):
+                adapter_v2_linear_with_bias_and_scale(module)
```
Comment (awaelchli) [inline finetune/adapter_v2.py:102]: Could we put these 3 lines into a utility function, or perhaps combine into `adapter_v2_linear_with_bias_and_scale` itself, what do you think? (we would have to do it in all files)

File: finetune/adapter_v2.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+
+    adapter_v2_mark_only_adapter_as_trainable(model)
```
Comment (awaelchli) [inline finetune/adapter_v2.py:105]: Perhaps the name could be shortened a bit by avoiding the name "adapter" twice
```suggestion
   mark_only_adapter_v2_as_trainable(model)
```

File: lit_llama/adapter_v2.py (hunk 1 above)
Comment (awaelchli) [reply lit_llama/adapter_v2.py:19]: (and there is a left-over debug print)

**Review by awaelchli — COMMENTED** [review]
@rasbt Thanks a lot for this. As always, I'm very impressed by your quick turnaround time! 🚀 I left minor suggestions for the utility functions

File: lit_llama/adapter_v2.py (hunk 1 above)
Comment (rasbt) [reply lit_llama/adapter_v2.py]: Thanks, this should be addressed now. I refactored this into a function and broke it down for readability purposes in the future (which may be overkill). It's now

```python
def get_adapter_substrings():
    substrings = ["adapter_wte", "gating_factor"]  # regular adapter v1 parameters
    substrings.extend(["adapter_scale", "adapter_bias"])  # adapter v2: new bias and scale used in Linear
    substrings.extend(["rms_1", "rms_2", "ln_f"])  # adapter v2: RMSNorm parameters are now trainable
    return substrings
```

File: finetune/adapter_v2.py (hunk 2 above)
Comment (rasbt) [reply finetune/adapter_v2.py]: Good call, I made this a function.

File: finetune/adapter_v2.py (hunk 3 above)
Comment (rasbt) [reply finetune/adapter_v2.py]: I like it! Done!

File: evaluate/adapter_v2.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    else:
+        raise ValueError("invalid dataset name (wikitext, ptb, c4 are allowed)")
+    return testdata
+
+
+def main(
+    datasets: str = "wikitext,ptb,c4",
+    *,
+    # compilation fails as it does not support torch.complex64 for RoPE
+    # compile: bool = False,
```
Comment (carmocca) [inline evaluate/adapter_v2.py:53]: This comment is not true anymore. I would remove this.

```suggestion
```

File: evaluate/adapter_v2.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    checkpoint_path: Path = Path("checkpoints/lit-llama/7B/lit-llama.pth"),
+    tokenizer_path: Path = Path("checkpoints/lit-llama/tokenizer.model"),
+    dtype: str = "float32",
+    quantize: Optional[str] = None,
+) -> None:
+    """Generates text samples based on a pre-trained LLaMA model and tokenizer.
+
+    Args:
+        datasets: The datasets to use as a comma separated string
+        # compile: Whether to compile the model.
```
Comment (carmocca) [inline evaluate/adapter_v2.py:65]:
```suggestion
```

File: evaluate/adapter_v2.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        accelerator: The hardware to run on. Possible choices are:
+            ``"cpu"``, ``"cuda"``, ``"mps"``, ``"gpu"``, ``"tpu"``, ``"auto"``.
+        adapter_path: Path to the checkpoint with trained adapter weights, which are the output of
+            `finetune_adapter_v2.py`.
+        checkpoint_path: The checkpoint path to load.
+        tokenizer_path: The tokenizer path to load.
+        quantize: Whether to quantize the model and using which method:
```
Comment (carmocca) [inline evaluate/adapter_v2.py:72]: dtype is missing from the docstring

File: evaluate/adapter_v2.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        raise ValueError(f"{dtype} is not a valid dtype.")
+    dtype = dt
+
+    print("Loading model ...", file=sys.stderr)
+    t0 = time.time()
+    with lazy_load(checkpoint_path) as pretrained_checkpoint, lazy_load(adapter_path) as adapter_checkpoint:
+        name = llama_model_lookup(pretrained_checkpoint)
+
+        with EmptyInitOnDevice(
+                device=fabric.device, dtype=dtype, quantization_mode=quantize
```
Comment (carmocca) [inline evaluate/adapter_v2.py:93]: you are using 4 spaces of indentation everywhere else

```suggestion
            device=fabric.device, dtype=dtype, quantization_mode=quantize
```

File: evaluate/adapter_v2.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+            test_string, bos=True, eos=False, device=fabric.device
+        )
+        encoded_text = encoded_text[
+            None, : 256 * model.config.block_size
+        ]  # add batch dimension, trim like gptq implementation
+        t0 = time.perf_counter()
+
+        nlls = 0
+        toks = 0
+        with torch.inference_mode():
```
Comment (carmocca) [inline evaluate/adapter_v2.py:131]: Why not wrap the entire function?

File: finetune/adapter_v2.py [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+                print(f"Saving adapter weights to {out_dir}")
+                # TODO: Provide a function/script to merge the adapter weights with pretrained weights
+                save_model_checkpoint(fabric, model, os.path.join(out_dir, f"iter-{iter_num:06d}.pth"))
+
+        dt = time.time() - t0
+        if iter_num % log_interval == 0:
+            fabric.print(f"iter {iter_num}: loss {loss.item():.4f}, time: {dt*1000:.2f}ms")
+
+
+def generate_response(model, instruction, input=""):
```
Comment (carmocca) [inline finetune/adapter_v2.py:164]: Should we try to import the functions that are the same from adapter.py?

This question applies to all these files. For instance, generate adapter does it: https://github.com/Lightning-AI/lit-llama/blob/main/generate/adapter.py#L14

File: howto/finetune_adapter_v2.md [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+We are able to demonstrate instruction-finetuning Lit-LLaMA 7B on the [Alpaca](https://github.com/tatsu-lab/stanford_alpaca) dataset on a **single GTX 3090 (24GB) GPU**. If using 8 GPUs, finetuning can be completed in under 1 hour.
+
+If you are new to LLaMA-Adapter and are interested to learn more about how it works before proceeding with the finetuning guide below, you might find our article [Understanding Parameter-Efficient Finetuning of Large Language Models: From Prefix Tuning to LLaMA-Adapters](https://lightning.ai/pages/community/article/understanding-llama-adapters/) helpful.
+
+LLaMA-Adapter v2 extends the original LLaMA-Adapter idea by adding trainable bias and scale parameters to each linear layer in the transformer. Furthermore, LLaMA-Adapter v2 makes the normalization layers trainable. Where the 7B LLaMA model has 1.2M trainable parameters with LLaMA v1, LLaMA-Adapter v2 adds an additional 2.8 M trainable parameter for the bias and scale parameters and ~300k trainable parameters for the normalization layers. So, adapter v2 has ~4.3 M trainable parameters in total.
```
Comment (carmocca) [inline howto/finetune_adapter_v2.md:9]: When would the user choose v1 over v2?

If the answer is never, should we replace the v1 implementation?
Otherwise, this should be explained in this howto

Comment (rasbt) [reply howto/finetune_adapter_v2.md:9]: I'd say that it makes sense to have both methods because Adapter v2 is more heavy-weight than v1. E.g. where v1 has 1.2 M trainable parameters, v2 has 4.3 M trainable parameters, which is almost 4x as much.

The parameter number is mentioned in the v2 how-to but not in the context of using v1 vs v2. I am happy to add that. 

Do you by chance have any additional ideas that come to mind regarding which version to use?

Comment (rasbt) [reply howto/finetune_adapter_v2.md:9]: One thing to highlight in the how-to is that the number of trainable parameters changes to 4.4, but inference cost doesn't change that much since using bias & scale parameters in the forward method are cheap, and the RMSNorm parameters (300k) were already there anyways.

Comment (carmocca) [reply howto/finetune_adapter_v2.md:9]: Good points, we should add those.

I haven't dove deep into the paper, but it describes v2 as an improved method for both multi-model data and text-only data:

> The newly designed framework also
exhibits stronger language-only instruction-following capabilities and even excels in chat interactions

File: finetune/adapter_v2.py (hunk 9 above)
Comment (awaelchli) [reply finetune/adapter_v2.py:164]: The important thing is that the script is kept as standalone as possible. So if someone wants to copy just the finetune script v2, they shouldn't need copy many other files besides just the model definition.

File: evaluate/adapter_v2.py (hunk 6 above)
Comment (rasbt) [reply evaluate/adapter_v2.py:70]: @carmocca I don't think this would make a difference in practice here, but it would look more organized. Changed it as you suggested!

File: finetune/adapter_v2.py (hunk 9 above)
Comment (rasbt) [reply finetune/adapter_v2.py:164]: should we keep it as this then? (And perhaps revisit this in case there is an adapter v3 some day 😅)

File: evaluate/adapter_v2.py (hunk 8 above)
Comment (rasbt) [reply evaluate/adapter_v2.py]: @carmocca ah yes, that makes more sense. Added the decorator to the orig adapter as well.

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
@carmocca let's merge it?
